pkg/kapp/diff/change_set_with_versioned_rs.go

```diff
@@ -76,6 +86,72 @@ func (d ChangeSetWithVersionedRs) Calculate() ([]Change, error) {
 	return allChanges, nil
 }
 
+func (d ChangeSetWithVersionedRs) checkForMaxDurationAnn(existingRsGrouped map[string][]ctlres.Resource,
+	existingNonVerRsGrouped map[string]ctlres.Resource) error {
+
+	for _, res := range d.newRs {
+		if val, found := res.Annotations()[maxDurationAnnKey]; found {
+
+			// Validate value of annotation `kapp.k14s.io/max-duration`.
+			duration, err := time.ParseDuration(val)
+			if err != nil {
+				return fmt.Errorf("%s for resource: %s", err.Error(), res.Description())
+			}
+
+			resKey := VersionedResource{res, nil}.UniqVersionedKey().String()
+			if exRes, found := existingNonVerRsGrouped[resKey]; found {
+				err = updateLastRenewedTime(res, exRes, duration)
+				if err != nil {
+					return err
+				}
+			} else if exRes, found := existingRsGrouped[resKey]; found {
+				// exRes contains list of all versioned resources currently present on the cluster for particular resource and are sorted
+				// here we will be using last created resource.
+				err = updateLastRenewedTime(res, exRes[len(exRes)-1], duration)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func updateLastRenewedTime(res, exRes ctlres.Resource, duration time.Duration) error {
+	var (
+		lastRenewed time.Time
+		err         error
+	)
+
+	val := exRes.Annotations()[lastRenewTimeAnnKey]
```

Comment:
Maybe we can extract this in `checkForDuration` itself? Then we wont have to pass on the exRes.